src/librustc/util/ppaux.rs

```diff
@@ -32,310 +33,564 @@ use syntax::ast::CRATE_NODE_ID;
+    ( $f:expr, $cx:expr $(, $kind:ident $data:tt)+ ) => {
+        Ok(())$(.and_then(|_| print_inner!($f, $cx, $kind $data)))+
+    };
 }
 
-fn fn_sig(f: &mut fmt::Formatter,
-          inputs: &[Ty],
-          variadic: bool,
-          output: Ty)
-          -> fmt::Result {
-    write!(f, "(")?;
-    let mut inputs = inputs.iter();
-    if let Some(&ty) = inputs.next() {
-        write!(f, "{}", ty)?;
-        for &ty in inputs {
-            write!(f, ", {}", ty)?;
-        }
-        if variadic {
-            write!(f, ", ...")?;
+
+struct LateBoundRegionNameCollector(Option<FxHashSet<Symbol>>);
+impl<'tcx> ty::fold::TypeVisitor<'tcx> for LateBoundRegionNameCollector {
+    fn visit_region(&mut self, r: ty::Region<'tcx>) -> bool {
+        match *r {
+            ty::ReLateBound(_, ty::BrNamed(_, name)) => {
+                if self.0.is_none() {
+                    self.0 = Some(FxHashSet());
+                }
+                self.0.as_mut().unwrap().insert(name);
+            },
+            _ => {},
         }
+        r.super_visit_with(self)
     }
-    write!(f, ")")?;
-    if !output.is_nil() {
-        write!(f, " -> {}", output)?;
+}
+
+#[derive(Debug)]
+pub struct PrintContext {
+    is_debug: bool,
+    is_verbose: bool,
+    identify_regions: bool,
+    used_region_names: Option<FxHashSet<Symbol>>,
+    region_index: usize,
+}
+impl Default for PrintContext {
+    fn default() -> Self {
+        ty::tls::with_opt(|tcx| {
+            let (is_verbose, identify_regions) = tcx.map(
+                |tcx| (tcx.sess.verbose(), tcx.sess.opts.debugging_opts.identify_regions)
+            ).unwrap_or((false, false));
+            PrintContext {
+                is_debug: false,
+                is_verbose: is_verbose,
+                identify_regions: identify_regions,
+                used_region_names: None,
+                region_index: 0,
+            }
+        })
     }
+}
+impl PrintContext {
+    fn is_verbose(&self) -> bool {
+        self.is_debug || self.is_verbose
+    }
+}
 
-    Ok(())
+trait DefaultWith<T> : Default {
+    fn default_with(_: &T) -> Self {
+        Default::default()
+    }
 }
 
-pub fn parameterized(f: &mut fmt::Formatter,
-                     substs: &subst::Substs,
-                     mut did: DefId,
-                     projections: &[ty::ProjectionPredicate])
-                     -> fmt::Result {
-    let key = ty::tls::with(|tcx| tcx.def_key(did));
-    let mut item_name = if let Some(name) = key.disambiguated_data.data.get_opt_name() {
-        Some(name)
-    } else {
-        did.index = key.parent.unwrap_or_else(
-            || bug!("finding type for {:?}, encountered def-id {:?} with no parent",
-                    did, did));
-        parameterized(f, substs, did, projections)?;
-        return write!(f, "::{}", key.disambiguated_data.data.as_interned_str());
-    };
+default impl<T> DefaultWith<T> for PrintContext {}
+
+impl<'tcx, T> DefaultWith<T> for PrintContext
```

Comment:
that's an... ok use of specialization. I suppose we can avoid it if we have to?

Reply:
Oops, I missed your review -- what do you mean by that? I can't quite get it...